From f87d54e6eede34a6ef4b77df56ec1df9e313822d Mon Sep 17 00:00:00 2001
From: Petr Lautrbach <lautrbach@redhat.com>
Date: Fri, 6 Feb 2026 14:31:52 +0100
Subject: [PATCH] Fix 32-bit ELF dynamic section parsing
Content-type: text/plain

Use Elf32_Dyn instead of Elf64_Dyn when parsing PT_DYNAMIC segments in
32-bit ELF files. The incorrect type could cause a segfault on special
32 bit libraries.

Fixes: https://issues.redhat.com/browse/RHEL-1357

Signed-off-by: Petr Lautrbach <lautrbach@redhat.com>
---
 src/library/file.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/library/file.c b/src/library/file.c
index 802a89cc028a..2dffaaceea1e 100644
--- a/src/library/file.c
+++ b/src/library/file.c
@@ -1485,7 +1485,7 @@ uint32_t gather_elf(int fd, off_t size)
 				if (ph_tbl[i].p_filesz > size)
 					goto err_out32;
 
-				Elf64_Dyn *dyn_tbl = malloc(ph_tbl[i].p_filesz);
+				Elf32_Dyn *dyn_tbl = malloc(ph_tbl[i].p_filesz);
 
 				if((unsigned int)lseek(fd, ph_tbl[i].p_offset,
 							SEEK_SET) !=
@@ -1494,7 +1494,7 @@ uint32_t gather_elf(int fd, off_t size)
 					goto err_out32;
 				}
 
-				num = ph_tbl[i].p_filesz / sizeof(Elf64_Dyn);
+				num = ph_tbl[i].p_filesz / sizeof(Elf32_Dyn);
 				if (num > 1000) {
 					free(dyn_tbl);
 					goto err_out32;
-- 
2.53.0

